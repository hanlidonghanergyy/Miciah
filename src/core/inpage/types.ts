export enum JsonRpcMethod {
  // synchronous or asynchronous
  eth_chainId = "eth_chainId",
  net_version = "net_version",
  eth_accounts = "eth_accounts",
  eth_coinbase = "eth_coinbase",
  eth_uninstallFilter = "eth_uninstallFilter", // synchronous

  // asynchronous only
  eth_requestAccounts = "eth_requestAccounts",
  eth_sign = "eth_sign",
  eth_ecRecover = "eth_ecRecover",
  personal_sign = "personal_sign",
  personal_ecRecover = "personal_ecRecover",
  eth_signTransaction = "eth_signTransaction",
  eth_sendRawTransaction = "eth_sendRawTransaction",
  eth_sendTransaction = "eth_sendTransaction",
  eth_signTypedData = "eth_signTypedData",
  eth_signTypedData_v1 = "eth_signTypedData_v1",
  eth_signTypedData_v2 = "eth_signTypedData_v2",
  eth_signTypedData_v3 = "eth_signTypedData_v3",
  eth_signTypedData_v4 = "eth_signTypedData_v4",
  wallet_getPermissions = "wallet_getPermissions",
  wallet_requestPermissions = "wallet_requestPermissions",
  wallet_addEthereumChain = "wallet_addEthereumChain",
  wallet_switchEthereumChain = "wallet_switchEthereumChain",
  wallet_watchAsset = "wallet_watchAsset",

  // asynchronous pub/sub
  eth_subscribe = "eth_subscribe",
  eth_unsubscribe = "eth_unsubscribe",

  // asynchronous filter methods
  eth_newFilter = "eth_newFilter",
  eth_newBlockFilter = "eth_newBlockFilter",
  eth_newPendingTransactionFilter = "eth_newPendingTransactionFilter",
  eth_getFilterChanges = "eth_getFilterChanges",
  eth_getFilterLogs = "eth_getFilterLogs",
}

export const AUTHORIZED_METHODS = new Set<string>([
  JsonRpcMethod.eth_sign,
  JsonRpcMethod.personal_sign,
  JsonRpcMethod.eth_signTransaction,
  JsonRpcMethod.eth_signTypedData,
  JsonRpcMethod.eth_signTypedData_v1,
  JsonRpcMethod.eth_signTypedData_v2,
  JsonRpcMethod.eth_signTypedData_v3,
  JsonRpcMethod.eth_signTypedData_v4,
]);

/**
 * A String specifying the version of the JSON-RPC protocol.
 * MUST be exactly "2.0".
 */
export type JsonRpcVersion = "2.0";

/**
 * An identifier established by the Client that MUST contain a String, Number,
 * or NULL value if included. If it is not included it is assumed to be a
 * notification. The value SHOULD normally not be Null and Numbers SHOULD
 * NOT contain fractional parts.
 */
export type JsonRpcId = number | string | null;

export interface JsonRpcError {
  code: number;
  message: string;
  data?: unknown;
  stack?: string;
}

export interface JsonRpcRequest<T> {
  jsonrpc: JsonRpcVersion;
  method: string;
  id: JsonRpcId;
  params?: T;
}

export interface JsonRpcNotification<T> {
  jsonrpc: JsonRpcVersion;
  method: string;
  params?: T;
}

export type JsonRpcCallback<T> = (
  error: Error | null,
  result?: JsonRpcResponse<T>
) => void;

export type JsonRpcCallbackBatch = (
  error: Error | null,
  result?: JsonRpcResponse<unknown>[]
) => void;

interface JsonRpcResponseBase {
  jsonrpc: JsonRpcVersion;
  id: JsonRpcId;
}

export interface JsonRpcSuccess<T> extends JsonRpcResponseBase {
  result: T;
}

export interface JsonRpcFailure extends JsonRpcResponseBase {
  error: JsonRpcError;
}

export type JsonRpcResponse<T> = JsonRpcSuccess<T> | JsonRpcFailure;
