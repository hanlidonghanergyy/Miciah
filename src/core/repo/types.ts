import { AccountParams } from "core/types";

export enum RepoTable {
  Networks = "networks",
  Accounts = "accounts",
}

export interface INetwork {
  chainId: number;
  rpcUrls: string[];
  name: string;
  mainnet: boolean;
  nativeCurrency: {
    name: string;
    symbol: string; // 2-6 characters long
    decimals: 18;
  };
  blockExplorerUrls?: string[];
  iconUrls?: string[];
}

export type IAccount = AccountParams & {
  address: string;
  name: string;
  usdValues: Record<number, string>;
};
