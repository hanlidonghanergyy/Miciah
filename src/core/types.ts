export enum TakyMessageType {
  PageMessage = "PAGE_MESSAGE",
}

export interface TakyMessageBase {
  type: TakyMessageType;
}

export interface TakyPermissionMessage extends TakyMessageBase {
  type: TakyMessageType.PageMessage;
}
