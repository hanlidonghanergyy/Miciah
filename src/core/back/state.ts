import create from "zustand/vanilla";
import { assert } from "lib/system/assert";
import { WalletStatus } from "core/types";
import { Vault } from "./vault";

export type State = {
  status: WalletStatus;
  vault: Vault | null;
  init: (vaultExists: boolean) => void;
  unlock: (vault: Vault) => void;
  lock: () => void;
};

export type UnlockedState = State & {
  status: WalletStatus.Ready;
  vault: Vault;
};

export type NotReadyState = State & {
  status: WalletStatus.Idle | WalletStatus.Welcome | WalletStatus.Locked;
  vault: null;
};

const store = create<State>((set) => ({
  status: WalletStatus.Idle,
  vault: null,
  init: (vaultExists) =>
    set({
      status: vaultExists ? WalletStatus.Locked : WalletStatus.Welcome,
      vault: null,
    }),
  unlock: (vault) => set({ status: WalletStatus.Ready, vault }),
  lock: () =>
    set((s) => ({
      status: s.status === WalletStatus.Ready ? WalletStatus.Locked : s.status,
      vault: null,
    })),
}));

export async function ensureInited() {
  const state = store.getState();
  if (state.status === WalletStatus.Idle) {
    const vaultExists = await Vault.isExist();
    state.init(vaultExists);
  }
}

export function getStatus() {
  const { status } = store.getState();
  return status;
}

export function withUnlocked<T>(factory: (state: UnlockedState) => T) {
  const state = store.getState();
  assertUnlocked(state);
  return factory(state);
}

export function withNotReady<T>(factory: (state: NotReadyState) => T) {
  const state = store.getState();
  assertNotReady(state);
  return factory(state);
}

export function assertUnlocked(state: State): asserts state is UnlockedState {
  assert(
    state.status === WalletStatus.Ready && state.vault instanceof Vault,
    "Wallet locked"
  );
}

export function assertNotReady(state: State): asserts state is NotReadyState {
  assert(state.status !== WalletStatus.Ready, "Disallowed for unlocked wallet");
}
