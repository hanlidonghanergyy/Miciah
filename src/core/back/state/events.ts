import { createEvent, merge } from "effector";

import { Approval, Account } from "core/types";

import { Vault } from "../vault";

export const inited = createEvent<boolean>();

export const unlocked = createEvent<{
  vault: Vault;
  accounts: Account[];
  hasSeedPhrase: boolean;
  approvals?: Approval[];
}>();

export const locked = createEvent();

export const seedPhraseAdded = createEvent();

export const accountsUpdated = createEvent<Account[]>();

export const walletPortsCountUpdated = createEvent<number>();

export const approvalAdded = createEvent<Approval>();

export const approvalResolved = createEvent<string>();

export const approvalsRejected = createEvent<string[] | null>();

export const approvalsChanged = merge([
  approvalAdded,
  approvalResolved,
  approvalsRejected,
]);

export const syncStarted = createEvent<number | string>();

export const synced = createEvent<number | string>();
