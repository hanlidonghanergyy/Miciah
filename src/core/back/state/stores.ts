import { createStore, combine } from "effector";

import { WalletStatus, Approval, Account, ActivityType } from "core/types";
import { getPageOrigin } from "core/common/permissions";

import { Vault } from "../vault";
import {
  inited,
  unlocked,
  locked,
  seedPhraseAdded,
  accountsUpdated,
  walletPortsCountUpdated,
  approvalAdded,
  approvalResolved,
  syncStarted,
  synced,
} from "./events";

export const $walletStatus = createStore(WalletStatus.Idle)
  .on(inited, (_s, vaultExists) =>
    vaultExists ? WalletStatus.Locked : WalletStatus.Welcome
  )
  .on(unlocked, () => WalletStatus.Unlocked)
  .on(locked, (state) =>
    state === WalletStatus.Unlocked ? WalletStatus.Locked : state
  );

export const $hasSeedPhrase = createStore(false)
  .on(unlocked, (_s, { hasSeedPhrase }) => hasSeedPhrase)
  .on(seedPhraseAdded, () => true)
  .on(locked, () => false);

export const $walletState = combine([$walletStatus, $hasSeedPhrase]);

export const $accounts = createStore<Account[]>([])
  .on(unlocked, (_s, { accounts }) => accounts)
  .on(accountsUpdated, (_s, accounts) => accounts)
  .on(locked, () => []);

export const $vault = createStore<Vault | null>(null)
  .on(unlocked, (_s, { vault }) => vault)
  .on(locked, (prevVault) => {
    prevVault?.cleanup();

    return null;
  });

export const $autoLockTimeout = createStore<MaybeTimeout>(null)
  .on(locked, (t) => {
    if (t !== null) clearTimeout(t);
    return null;
  })
  .on(walletPortsCountUpdated, (t, count) => {
    if (t !== null) clearTimeout(t);

    const timeout = count === 0 && 0; /*Global.get(Setting.AutoLockTimeout)*/
    return timeout ? setTimeout(() => locked(), +timeout) : null;
  });

type MaybeTimeout = ReturnType<typeof setTimeout> | null;

export const $syncPool = createStore<number[]>([])
  .on(syncStarted, (pool, chainId) => [...pool, chainId])
  .on(synced, (pool, chainId) => {
    const i = pool.indexOf(chainId);

    if (i > -1) {
      // Remove one element by index
      const copied = Array.from(pool);
      copied.splice(i, 1);
      return copied;
    }

    return pool;
  });

export const $syncStatus = $syncPool.map((pool) => Array.from(new Set(pool)));

export const $approvals = createStore<Approval[]>([])
  .on(approvalAdded, (approvals, newApproval) => {
    if (newApproval.type === ActivityType.Connection) {
      const newApprovalOrigin = getPageOrigin(newApproval.source);
      for (const approval of approvals) {
        if (
          approval.type === ActivityType.Connection &&
          getPageOrigin(approval.source) === newApprovalOrigin
        ) {
          return;
        }
      }
    }

    return [...approvals, newApproval];
  })
  .on(approvalResolved, (current, approvalId) =>
    current.filter((a) => a.id !== approvalId)
  )
  .on(locked, () => []);
