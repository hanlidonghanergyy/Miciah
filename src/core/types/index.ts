export * from "./base";
export * from "./storage";
export * from "./account";
export * from "./network";
export * from "./messages";
export * from "./shared";
export * from "./rpc";
export * from "./activity";
export * from "./tokens";
export * from "./sync";
export * from "./fees";
export * from "./contacts";
export * from "./permissions";
