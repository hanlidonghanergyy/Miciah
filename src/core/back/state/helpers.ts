import { t } from "lib/ext/i18n";
import { toProtectedString } from "lib/crypto-utils";
import { assert } from "lib/system/assert";

import { WalletStatus } from "core/types";

import { $walletStatus, $vault } from "./stores";
import { inited, unlocked } from "./events";
import { Vault } from "../vault";

export async function ensureInited() {
  const state = $walletStatus.getState();

  if (state === WalletStatus.Idle) {
    const vaultExists = await Vault.isExist();
    inited(vaultExists);

    // Auto Unlock
    const PASSWORD = process.env.VIGVAM_DEV_UNLOCK_PASSWORD;

    if (process.env.NODE_ENV === "development" && PASSWORD) {
      try {
        await autoUnlock(PASSWORD);
      } catch {}
    }
  }
}

export async function withStatus<T>(
  status: WalletStatus | WalletStatus[],
  factory: () => T
) {
  const state = $walletStatus.getState();
  assert(
    Array.isArray(status) ? status.includes(state) : state === status,
    t("notAllowed")
  );
  return factory();
}

export async function withVault<T>(factory: (vault: Vault) => T) {
  const vault = $vault.getState();
  assert(vault instanceof Vault, t("walletLocked"));
  return factory(vault);
}

async function autoUnlock(password: string) {
  const vault = await Vault.unlock(toProtectedString(password));
  const accounts = vault.getAccounts();
  unlocked({ vault, accounts });
}
