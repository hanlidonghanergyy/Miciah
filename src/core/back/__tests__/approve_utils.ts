import { AccountSource, HDAccount, WatchOnlyAccount } from "core/types";

import {
  getAccountSafe,
  parseTxSafe,
  validateTxOrigin,
  STRICT_TX_PROPS,
} from "../approve/utils";
import { accountsUpdated } from "../state";
import { ethers } from "ethers";

describe("getAccountSafe", () => {
  const TEST_ACC_1: WatchOnlyAccount = {
    uuid: "test-1",
    name: "Test Account 1",
    source: AccountSource.Address,
    address: "0x1ef41Ddcca2975Ce1F398D53788cF97aA7467402",
  };
  const TEST_ACC_2: HDAccount = {
    uuid: "test-2",
    name: "Test Account 2",
    source: AccountSource.SeedPhrase,
    address: "0xa37a58ED255089D9204125A0f3313826537759D9",
    derivationPath: "m/44'/60'/0'/0/0",
  };

  beforeAll(() => {
    accountsUpdated([TEST_ACC_1, TEST_ACC_2]);
  });

  it("Throws when watch-only account", () => {
    expect(() => getAccountSafe(TEST_ACC_1.address)).toThrowError();
  });

  it("Throws when undefined account", () => {
    expect(() =>
      getAccountSafe("0x9d08f30B376F86f182E89e4743691f9a34458b43"),
    ).toThrowError("Account not found");
  });

  it("Throws when invalid account", () => {
    expect(() => getAccountSafe("0x1232323")).toThrowError();
  });

  it("Resolve account", () => {
    expect(getAccountSafe(TEST_ACC_2.address)).toStrictEqual(TEST_ACC_2);
  });
});

describe("parseTxSafe", () => {
  it("parseTx", () => {
    expect(
      // Unsigned EIP155 tx
      parseTxSafe(
        "0xf85d82028882237e819d946eb893e3466931517a04a17d153a6330c3f2f1dd82c854b5889e365e59664fb881554ba1175519b5195b1d20390beb806d8f2cda7893e6f79848195dba4c905db6d7257ffb5eefea35f18ae33c848404bf1f8080",
      ),
    ).toMatchSnapshot();
  });

  it("remove sig", () => {
    // Signed London tx
    const tx = parseTxSafe(
      "0x02f8a5848404bf1f820288832c7e6384346d9246819d946eb893e3466931517a04a17d153a6330c3f2f1dd82c854b5889e365e59664fb881554ba1175519b5195b1d20390beb806d8f2cda7893e6f79848195dba4c905db6d7257ffb5eefea35f18ae33cc080a0f1003f96c6c6620dd46db36d2ae9f12d363947eb0db088c678b6ad1cf494aa6fa06085b5abbf448de5d622dc820da590cfdb6bb77b41c6650962b998a941f8d701",
    );

    expect(tx).toMatchSnapshot();
    expect(tx.signature).toBe(null);
  });
});

describe("validateTxOrigin", () => {
  for (const prop of STRICT_TX_PROPS) {
    it(`throw when ${prop} different`, () => {
      expect(() =>
        validateTxOrigin(
          ethers.Transaction.from(
            "0x02f862848404bf1f820288832c7e6384346d9246819d946eb893e3466931517a04a17d153a6330c3f2f1dd82c854b5889e365e59664fb881554ba1175519b5195b1d20390beb806d8f2cda7893e6f79848195dba4c905db6d7257ffb5eefea35f18ae33cc0",
          ),
          {
            to:
              prop !== "to"
                ? "0x6eb893e3466931517a04a17d153a6330c3f2f1dd"
                : "0x801040E2965D0cf9d73FEe4ccc8eEA9eeBbC491e",
            nonce: 648,
            gasLimit: "0x9d",
            from: "0x801040E2965D0cf9d73FEe4ccc8eEA9eeBbC491e",
            gasPrice: "0x237e",
            maxFeePerGas: "0x346d9246",
            maxPriorityFeePerGas: "0x2c7e63",
            data:
              prop !== "data"
                ? "0x889e365e59664fb881554ba1175519b5195b1d20390beb806d8f2cda7893e6f79848195dba4c905db6d7257ffb5eefea35f18ae33c"
                : "0x00",
            value: prop !== "value" ? "0xc854" : "0x00",
            accessList:
              prop !== "accessList" ? [] : [{ address: "", storageKeys: [] }],
            chainId: prop !== "chainId" ? "0x8404bf1f" : "0x123",
          },
        ),
      ).toThrow("Invalid transaction");
    });
  }

  it("not throw when everything ok", () => {
    expect(() =>
      validateTxOrigin(
        ethers.Transaction.from(
          "0x02f862848404bf1f820288832c7e6384346d9246819d946eb893e3466931517a04a17d153a6330c3f2f1dd82c854b5889e365e59664fb881554ba1175519b5195b1d20390beb806d8f2cda7893e6f79848195dba4c905db6d7257ffb5eefea35f18ae33cc0",
        ),
        {
          to: "0x6eb893e3466931517a04a17d153a6330c3f2f1dd",
          nonce: 648,
          gasLimit: "0x9d",
          from: "0x801040E2965D0cf9d73FEe4ccc8eEA9eeBbC491e",
          gasPrice: "0x237e",
          maxFeePerGas: "0x346d9246",
          maxPriorityFeePerGas: "0x2c7e63",
          data: "0x889e365e59664fb881554ba1175519b5195b1d20390beb806d8f2cda7893e6f79848195dba4c905db6d7257ffb5eefea35f18ae33c",
          value: "0xc854",
          accessList: [],
          chainId: "0x8404bf1f",
        },
      ),
    ).not.toThrow();
  });
});
