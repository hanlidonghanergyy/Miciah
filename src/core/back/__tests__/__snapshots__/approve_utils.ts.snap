// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTxSafe parseTx 1`] = `
{
  "accessList": null,
  "chainId": "2214903583",
  "data": "0x889e365e59664fb881554ba1175519b5195b1d20390beb806d8f2cda7893e6f79848195dba4c905db6d7257ffb5eefea35f18ae33c",
  "gasLimit": "157",
  "gasPrice": "9086",
  "maxFeePerGas": null,
  "maxPriorityFeePerGas": null,
  "nonce": 648,
  "sig": null,
  "to": "0x6Eb893e3466931517a04a17D153a6330c3f2f1dD",
  "type": 0,
  "value": "51284",
}
`;

exports[`parseTxSafe remove sig 1`] = `
{
  "accessList": [],
  "chainId": "2214903583",
  "data": "0x889e365e59664fb881554ba1175519b5195b1d20390beb806d8f2cda7893e6f79848195dba4c905db6d7257ffb5eefea35f18ae33c",
  "gasLimit": "157",
  "gasPrice": null,
  "maxFeePerGas": "879596102",
  "maxPriorityFeePerGas": "2915939",
  "nonce": 648,
  "sig": null,
  "to": "0x6Eb893e3466931517a04a17D153a6330c3f2f1dD",
  "type": 2,
  "value": "51284",
}
`;
